sdk/eventhub/Azure.Messaging.EventHubs/src/EventProcessorClient.cs

```diff
@@ -287,194 +437,404 @@ public class EventProcessorClient : EventProcessorBase<PartitionContext>, IAsync
             FullyQualifiedNamespace = fullyQualifiedNamespace;
             EventHubName = eventHubName;
             ConsumerGroup = consumerGroup;
             Manager = partitionManager;
-            Options = clientOptions;
+            ClientOptions = clientOptions;
             RetryPolicy = clientOptions.RetryOptions.ToRetryPolicy();
-            Identifier = Guid.NewGuid().ToString();
+            Identifier = string.IsNullOrEmpty(clientOptions.Identifier) ? Guid.NewGuid().ToString() : clientOptions.Identifier;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="EventProcessorClient"/> class.
+        /// </summary>
+        ///
+        protected EventProcessorClient()
+        {
         }
 
         /// <summary>
         ///   Initializes a new instance of the <see cref="EventProcessorClient"/> class.
         /// </summary>
         ///
-        /// <param name="consumerGroup">The name of the consumer group this processor is associated with.  Events are read in the context of this group.</param>
         /// <param name="partitionManager">Interacts with the storage system with responsibility for creation of checkpoints and for ownership claim.</param>
+        /// <param name="consumerGroup">The name of the consumer group this processor is associated with.  Events are read in the context of this group.</param>
         /// <param name="connection">The <see cref="EventHubConnection" /> connection to use for communication with the Event Hubs service.</param>
         /// <param name="clientOptions">The set of options to use for this processor.</param>
         ///
         /// <remarks>
         ///   This constructor is intended only to support functional testing and mocking; it should not be used for production scenarios.
         /// </remarks>
         ///
-        protected internal EventProcessorClient(string consumerGroup,
-                                                PartitionManager partitionManager,
-                                                EventHubConnection connection,
-                                                EventProcessorClientOptions clientOptions)
+        internal EventProcessorClient(PartitionManager partitionManager,
+                                      string consumerGroup,
+                                      EventHubConnection connection,
+                                      EventProcessorClientOptions clientOptions)
         {
-            // TODO: we probably can remove this constructor and OwnsConnection property because the processor does not have
-            // a single connection anymore.  In fact, returning the same connection from the factory might result in undefined
-            // behavior.  We could take a connection factory here instead of a single connection.
+            // TODO: we probably can remove this constructor because the processor does not have a single connection anymore.
```

Comment:
Great. I'll make this constructor into a mockable-factory one. Should we make it `protected` or `internal` only? I'll keep it as `internal` for now.

I'm not sure if we should make it `protected` now if we don't know for sure that we'll keep using the connection/consumer factory approach from now on. However, hiding it will make mocking the processor harder.

Reply:
Updated it. Included the namespace and the event hub name in the parameters because we don't have a single connection to get these names from anymore. It's in a separated commit, so if you think we should revert it, it should be simple (I think).

Reply:
I think that makes sense to me.